Implementation of service

KartaviK opened the pull request.

KartaviK committed:
Interface for service

KartaviK committed:
Add implementation of phonet authorization

KartaviK committed:
Implement Service

KartaviK committed:
More tests for service

Horat1us commented on src/Service.php:
```diff
@@ -0,0 +1,370 @@
+    protected $authProvider;
+
+    public function __construct(
+        GuzzleHttp\ClientInterface $client,
```
Change arguments to config -> provider -> client

KartaviK committed:
Separate Service into Service and Repository

KartaviK committed:
Add test for missed calls

KartaviK committed:
Remove getCompleteCallsRequest method

sashabeton commented on src/Authorization/CacheException.php:
```diff
@@ -0,0 +1,41 @@
+ * Class CacheException
+ * @package Wearesho\Phonet\Authorization
+ */
+class CacheException extends \RuntimeException
```
Override `toString()` to receive correct error descriptions

sashabeton commented on src/Authorization/CacheProvider.php:
```diff
@@ -0,0 +1,85 @@
+     *
+     * @throws \Psr\SimpleCache\InvalidArgumentException
+     */
+    protected function cacheResponse(string $cacheKey, GuzzleHttp\Cookie\CookieJarInterface $response)
```
Add return type for elegance

KartaviK committed:
Optimize provide method for cache provider
Add return type for protected method setCache
Override CacheException toString method for details

KartaviK committed:
Add test for repository companyCalls method
Refactor test for active call
Add test for complete call

KartaviK committed:
Refactor complete calls tests used phpunit dependency with dataProvider

KartaviK committed:
Fix tests

KartaviK committed:
Add test for users() method in repository
Add test for employee json serialize

sashabeton commented on src/Authorization/CacheProvider.php:
```diff
@@ -0,0 +1,78 @@
+
+        $cached = $this->cache->get($cacheKey);
+
+        if (!$cached instanceof GuzzleHttp\Cookie\CookieJarInterface || !$this->cache->has($cacheKey)) {
```
Remove useless second condition

KartaviK committed:
Remove useless condition for cache provider provide method

Horat1us commented on composer.json:
```diff
@@ -4,11 +4,17 @@
+        "horat1us/environment-config": "^1.3",
+        "myclabs/php-enum": "^1.6",
+        "nesbot/carbon": "^1.36",
+        "phpunit/phpunit": "^7.5",
```
Why not dev dependencies?

Horat1us commented on src/Config.php:
```diff
@@ -0,0 +1,28 @@
+    use ClientTrait;
+
+    public function __construct(
+        ClientInterface $client,
```
Should be used as GuzzleHttp\ClientInterface, because `ClientInterface` is too common.

Horat1us commented on src/Data/ActiveCall.php:
```diff
@@ -0,0 +1,101 @@
+        Enum\LastEvent $lastEvent,
+        Employee $employeeCaller,
+        ?Employee $employeeCallTaker,
+        ?Collection\Subject $subjects,
```
Необязательные аргументы в конец.

Horat1us commented on src/Model.php:
```diff
@@ -0,0 +1,111 @@
+     * @return ResponseInterface
+     * @throws GuzzleHttp\Exception\GuzzleException
+     */
+    protected function send(GuzzleHttp\Psr7\Request $request): ResponseInterface
```
А почему не вынести это в отдельный класс, который будет отправлять запросы? Тогда не нужны будут тесты для абстрактных классов, а тесты собственно репозитория и сервиса можно будет мокать без использования Guzzle. И зависимости авторизации/конфига как раз передавать в сервис, который уже отправляет запросы. Тогда Repository\Service будут заниматься формированием и парсингом запроса, которые у всех разные.

KartaviK committed:
Refactor architecture

KartaviK committed:
Documentation

KartaviK committed:
Fix api docs

KartaviK committed:
Few refactor of return type of sender method

KartaviK committed:
Add test for not json response body
